### ISSUE TITLE
Changing a column type between Integer and Float does not alter the default value, failing further migrations

### ISSUE THREAD

**Issue by kurtportelli** [piccolo-orm/piccolo#1071]
### Steps to Reproduce

1. Original table already in db schema:
```
class MyTable(Table):
    my_column = Integer()
```

2. Change the existing column from Integer to Float:
```
class MyTable(Table):
    my_column = Float()
```

3. Generate a migration file with `piccolo migrations new piccolo_app --auto`:
```
manager.alter_column(
    table_class_name="MyTable",
    tablename="my_table",
    column_name="my_column",
    db_column_name="my_column",
    params={},
    old_params={},
    column_class=Float,
    old_column_class=Integer,
    schema=None,
)
```

4. Optional (same result with or without doing this step) - Apply the migration `piccolo migrations forwards piccolo_app`

5. Optional (same result with or without doing this step) -  Make another change elsewhere in the db schema

6. Try to generate another migration file with `piccolo migrations new piccolo_app --auto`.  This fails with the following error:
```
🚀 Creating new migration ...
The command failed.
The default 0 isn't one of the permitted types - (<class 'float'>, None)
```

---

Changing from Float to Integer (vice versa of the above) results in a similar:
```
🚀 Creating new migration ...
The command failed.
The default 0.0 isn't one of the permitted types - (<class 'int'>, None)
```

---

### Workaround

1. Manually amend the following two lines in the first migration generated in step 3 above:
```
    params={"default": 0.0},
    old_params={"default": 0},
```

2. Or vice-versa if changing from Float to Integer.

3. Generating a new migration as per step 6 above now succeeds.

---

### Further Investigation

1. If the column in question has both `null=True` and `default=None` parameters this does not happen, despite the generated migration file being identical. This is because the default is supported for both the old column type (nullable Integer) and the new one (nullable Float).

2. If it only has `null=True`, the issue still occurs. This is because despite the column being nullable, the default was still 0 / 0.0

---

### Conclusion

Essentially, the generated migration alters the column, but not its default value. This fails any future migrations when the old default value is not valid for the new column type. The issue is not raised when generating the original migration nor when applying it. It is only evident when trying to generate any further migrations.

**Comment by sinisaos** [comment]
@kurtportelli Yes, you are right, I can reproduce a bug. I think this could be solved by adding another condition to the [_validate_default](https://github.com/piccolo-orm/piccolo/blob/716b7f05e48433f69b60e1cc5539f0713011751e/piccolo/columns/base.py#L514) method in the base `Column` class like this
```python
def _validate_default(
    self,
    default: t.Any,
    allowed_types: t.Iterable[t.Union[None, t.Type[t.Any]]],
    allow_recursion: bool = True,
) -> bool:
    ...
    # new condition
    elif (
        isinstance(default, (int, float))
        and type(default) in allowed_types
    ):
        self._validated = True
        return True

     raise ValueError(
            f"The default {default} isn't one of the permitted types - "
            f"{allowed_types}"
        )
```
but we need to wait @dantownsend  to see if that good solution.

**Comment by dantownsend** [comment]
It's an interesting bug. It might be because Piccolo's migrations try to detect any changes in the columns, and for the default value it just changes from `0` to `0.0`.

```python
>>> 0 == 0.0
True
```

I need to dig into it more to find out for sure.

@kurtportelli Thanks for reporting the issue, and @sinisaos for looking into it.

**Comment by dantownsend** [comment]
The workaround for now is to modify the `alter_column` in the migration file to this:

```python

manager.alter_column(
    params={
        "default": 0.0,
    },
    old_params={
        "default": 0,
    },
)
```

**Comment by dantownsend** [comment]
This is the problematic code:

https://github.com/piccolo-orm/piccolo/blob/47e530520e85d9c01c02a5ce62771ebc62adb00a/piccolo/apps/migrations/auto/diffable_table.py#L167-L170

Which calls this:

https://github.com/piccolo-orm/piccolo/blob/47e530520e85d9c01c02a5ce62771ebc62adb00a/piccolo/apps/migrations/auto/diffable_table.py#L19-L50

So we either modify that to also compare the types of the default args.

Or we modify `_validate_default` to be less strict, because `0` as a default for a `Float` column is fine really.

**Comment by dantownsend** [comment]
Having thought about it a bit more, I think we need to fix `compare_dicts` as that is the root cause of the problem.

**Comment by dantownsend** [comment]
This should be fixed in `piccolo==1.17.1`.

**Comment by kurtportelli** [comment]
That's great! Thanks for the quick fix!
